### PR TITLE
[multistage][refactor] Refactor RelNode to StageNode conversion

### PR DESCRIPTION
**Pull request by xiangfu0**
1. Make RelNode to StageNode 1:1 mapping
2. Add ExchangeNodeSplitterVisitor to make ExchangeNode to MailboxReceiveNode and MailboxSendNode
3. Add StageIdAssigner to go over plan and assign stageId

### CODE REVIEW

**Review by walterddr — APPROVED** [review]

**Review by walterddr — COMMENTED** [review]
let's unblock the refactoring and continue address comments on follow up PR.

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ExchangeNodeSplitterVisitor.java
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ExchangeNodeSplitterVisitor.java:41]: are we planning the make the exchange node splitter more generic going forward? cuz. 
1. if we were to support more complex logic such as pipeline breaker, then this API seems not enough to indicate that (as it only can split into MailboxSend/ReceiveNode; 
    - are we planning to support other type of split going forward? if so what would that be? such as subquery, etc)
2. if we decided to add another visitor to support other type of exchange that will not be converted into mailbox send/receive, then the logic should be merged with StageAssigner --> b/c the moment we do exchange split is the same moment we make decision to assign stage IDs. 

Let's make sure we are clear on what's the hierarch. us the example of subquery execution are we abstracting into:
1. convert exchange into subqueries and attach subqueryID
2. convert the rest of the exchanges into mailboxes + attach stageIDs; 
3. convert the rest of the exchanges into some other abstraction.

or
1. convert exchanges into subquery node and mailbox nodes (and some other nodes)
2. visit and assign subqueryID and stageID (and some other nodes)

Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ExchangeNodeSplitterVisitor.java:41]: subPlanExchange --> indicate subPlan boundary --> it's own planNode root will not be run (held on broker) until all children returned and info is being converted into its own plan

Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ExchangeNodeSplitterVisitor.java:41]: mailboxExchange --> indicate a planFragment boundary --> its own planRoot and all its children that will interact with it via mailbox; all planFragment will be dispatch from broker at the same time and will run on every server

Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ExchangeNodeSplitterVisitor.java:41]: pipelineBreakerExchange --> indicate part of the planFragment can wait on another part of the fragment --> this wait decision is made on server

Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ExchangeNodeSplitterVisitor.java]: Merged ExchangeNodeSplitterVisitor and StageIdAssigner to StageFragmenter which will look for *ExchangeNode* and split it to MailboxReceiveNode/MailboxSendNode pair.

Later on we can extend the logic here to introduce more types of *ExchangeNode*, like `SubPlanExchangeNode` or `PipelineBreakerExchangeNode`

**Pull request merged**
